### PR TITLE
LVM writecache support

### PR DESCRIPTION
**Pull request by vojtechtrefny**

### CODE REVIEW

File: src/lib/plugin_apis/lvm.api [hunk 1]
```diff
@@ -1447,6 +1447,54 @@ gchar* bd_lvm_cache_pool_name (const gchar *vg_name, const gchar *cached_lv, GEr
  */
 BDLVMCacheStats* bd_lvm_cache_stats (const gchar *vg_name, const gchar *cached_lv, GError **error);
 
+/**
+ * bd_lvm_writecache_attach:
+ * @vg_name: name of the VG containing the @data_lv and the @cache_pool_lv LVs
+ * @data_lv: data LV to attache the @cache_pool_lv to
```
Comment (tbzatek) [inline src/lib/plugin_apis/lvm.api:1453]: a typo

File: src/plugins/lvm-dbus.c [hunk 2]
```diff
@@ -3369,6 +3375,130 @@ gboolean bd_lvm_cache_create_cached_lv (const gchar *vg_name, const gchar *lv_na
     return TRUE;
 }
 
+/**
+ * bd_lvm_writecache_attach:
+ * @vg_name: name of the VG containing the @data_lv and the @cache_pool_lv LVs
+ * @data_lv: data LV to attache the @cache_pool_lv to
```
Comment (tbzatek) [inline src/plugins/lvm-dbus.c:3381]: the same typo.

File: src/plugins/lvm-dbus.c [hunk 3]
```diff
@@ -3369,6 +3375,130 @@ gboolean bd_lvm_cache_create_cached_lv (const gchar *vg_name, const gchar *lv_na
...
+ * Tech category: %BD_LVM_TECH_WRITECACHE-%BD_LVM_TECH_MODE_MODIFY
+ */
+gboolean bd_lvm_writecache_attach (const gchar *vg_name, const gchar *data_lv, const gchar *cache_lv, const BDExtraArg **extra, GError **error) {
+    GVariantBuilder builder;
+    GVariant *params = NULL;
+    gchar *lv_id = NULL;
+    gchar *lv_obj_path = NULL;
+    gboolean success = FALSE;
+
+    /* both LVs need to be inactive for the writecache convert to work */
```
Comment (tbzatek) [inline src/plugins/lvm-dbus.c:3398]: I think this fact is worth mentioning in the function docs.

File: src/plugins/lvm-dbus.c [hunk 4]
```diff
@@ -3369,6 +3375,130 @@ gboolean bd_lvm_cache_create_cached_lv (const gchar *vg_name, const gchar *lv_na
...
+        return FALSE;
+    g_variant_builder_init (&builder, G_VARIANT_TYPE_TUPLE);
+    g_variant_builder_add_value (&builder, g_variant_new ("o", lv_obj_path));
+    params = g_variant_builder_end (&builder);
+    g_variant_builder_clear (&builder);
+
+    lv_id = g_strdup_printf ("%s/%s", vg_name, cache_lv);
+
+    call_lvm_obj_method_sync (lv_id, LV_INTF, "WriteCacheLv", params, NULL, extra, TRUE, error);
+    return ((*error) == NULL);
```
Comment (tbzatek) [inline src/plugins/lvm-dbus.c:3420]: FYI, as a preparation for #527 I would suggest to add a guard (`g_warn_if_fail()` or similar) at the beginning of the function.

File: src/plugins/lvm-dbus.c [hunk 5]
```diff
@@ -3369,6 +3375,130 @@ gboolean bd_lvm_cache_create_cached_lv (const gchar *vg_name, const gchar *lv_na
...
+    if (!lv_obj_path)
+        return FALSE;
+    g_variant_builder_init (&builder, G_VARIANT_TYPE_TUPLE);
+    g_variant_builder_add_value (&builder, g_variant_new ("o", lv_obj_path));
+    params = g_variant_builder_end (&builder);
+    g_variant_builder_clear (&builder);
+
+    lv_id = g_strdup_printf ("%s/%s", vg_name, cache_lv);
+
+    call_lvm_obj_method_sync (lv_id, LV_INTF, "WriteCacheLv", params, NULL, extra, TRUE, error);
```
Comment (tbzatek) [inline src/plugins/lvm-dbus.c:3419]: Leaking `lv_id`?

File: src/plugins/lvm-dbus.c [hunk 6]
```diff
@@ -3369,6 +3375,130 @@ gboolean bd_lvm_cache_create_cached_lv (const gchar *vg_name, const gchar *lv_na
...
+    }
+
+    /* 1/3 steps done */
+    bd_utils_report_progress (progress_id, 33, "Cache LV created");
+
+    success = bd_lvm_lvcreate (vg_name, lv_name, data_size, NULL, slow_pvs, NULL, error);
+    if (!success) {
+        g_prefix_error (error, "Failed to create the data LV: ");
+        bd_utils_report_finished (progress_id, (*error)->message);
+        return FALSE;
```
Comment (tbzatek) [inline src/plugins/lvm-dbus.c:3483]: Leaking `name`

File: src/plugins/lvm.c [hunk 7]
```diff
@@ -2487,6 +2494,120 @@ gboolean bd_lvm_cache_create_cached_lv (const gchar *vg_name, const gchar *lv_na
...
+    }
+
+    /* 1/3 steps done */
+    bd_utils_report_progress (progress_id, 33, "Cache LV created");
+
+    success = bd_lvm_lvcreate (vg_name, lv_name, data_size, NULL, slow_pvs, NULL, error);
+    if (!success) {
+        g_prefix_error (error, "Failed to create the data LV: ");
+        bd_utils_report_finished (progress_id, (*error)->message);
+        return FALSE;
```
Comment (tbzatek) [inline src/plugins/lvm.c:2592]: Same as above - leaking `name`.

File: src/plugins/lvm-dbus.c (hunk 3 above)
Comment (vojtechtrefny) [reply src/plugins/lvm-dbus.c:3401]: Ok, I've added a note to the docstring.

File: src/plugins/lvm-dbus.c (hunk 5 above)
Comment (vojtechtrefny) [reply src/plugins/lvm-dbus.c:3422]: I should stop trusting code I copy-paste :-)

File: src/plugins/lvm-dbus.c (hunk 6 above)
Comment (vojtechtrefny) [reply src/plugins/lvm-dbus.c:3488]: Same here.

File: src/plugins/lvm.c (hunk 7 above)
Comment (vojtechtrefny) [reply src/plugins/lvm.c:2596]: And here too :-)

**Comment by vojtechtrefny** [comment]
This should be now ready for review. I've added a new commit with checking for LVM DBus API version. (The DBus API and lvm version with writecache support might change, I'll wait for LVM before merging this.)

File: src/plugins/lvm-dbus.c [hunk 8]
```diff
@@ -3369,6 +3375,134 @@ gboolean bd_lvm_cache_create_cached_lv (const gchar *vg_name, const gchar *lv_na
...
+    success = bd_lvm_lvdeactivate (vg_name, data_lv, NULL, error);
+    if (!success)
+        return FALSE;
+
+    success = bd_lvm_lvdeactivate (vg_name, cache_lv, NULL, error);
+    if (!success)
+        return FALSE;
+
+    lv_id = g_strdup_printf ("%s/%s", vg_name, data_lv);
+    lv_obj_path = get_object_path (lv_id, error);
```
Comment (tbzatek) [inline src/plugins/lvm-dbus.c:3414]: And seems like `get_object_path()` returns allocated string that needs to be freed.

Comment (vojtechtrefny) [reply src/plugins/lvm-dbus.c:3415]: Fixed.

**Pull request merged**